### ISSUE TITLE
Showing mini libs on left nav

### ISSUE THREAD

**Issue by bp74** [dart-lang/dartdoc#662]
The new API docs looks great, but i'm missing a feature i was using with the old tool. In my library i have a big library that users are referencing, which consists of smaller mini-libs (in fact they are not really mini, but i use them to group my classes).

Let's look at the output from the old tool:
http://www.stagexl.org/docs/api/index.html

On the left you see 13 sub-libraries which make up my big library. It much easier to learn the API with this grouping because you just see those classes that are interesting for a given context.

I have checked the options for the new dartdoc tool, but i couldn't find a way how to do this. Maybe i'm just missing something. Thanks for your help!

**Comment by sethladd** [comment]
Thanks for the feedback! Sounds like you'd like to make it easier to jump around libraries, from anywhere in dartdoc.

**Comment by bp74** [comment]
Well, it's not really about jumping around but more about grouping certain classes to make it more logical. For example if a user want's to learn more about the animation classes, he should only see the classes that are related to animation. Internally those classes are in the "stagexl.animation" library, but they are all exported in the "stagexl" library. 

In other languages they use namespaces for this. Related classes are in a common namespace and the documentation provides a navigation for this namespaces. I tried to replicate this with my libraries in StageXL and it worked pretty good with the old tool. 

Maybe there is an option in dartdoc that allows generating the docs like this, but i couldn't figure out how this could work.

**Comment by sethladd** [comment]
Any chance you could sketch out what you would like to see, scan it in, and attach it here?

> Maybe there is an option in dartdoc that allows generating the docs like this, but i couldn't figure out how this could work.

I don't think so. So you're not missing it :)

**Comment by bp74** [comment]
Sure i can sketch out what i'm talking about, but i think it's easy to see when you compare the old to new docs:

http://www.stagexl.org/docs/api/index.html (old)
http://www.stagexl.org/docs/api2/index.html (new)

The old docs show my internal libraries on the left (those are internal mini-libs which are combined to one big library that the user imports). Those internal mini-libs are used to get groups of classes just to get a structure to make it easier to explore the available classes. If the user clicks "stage.animation" he gets a list of the animation classes on the right. If the user clicks "stage.media" he get's a list of media classes on the right. 

Now let's look at the new docs. I can only export the "stagexl" library because this is the only library in the lib-folder. If i move my mini-libs to the lib-folder the dartdoc tool would probably do what i am asking for (showing the mini-libs on the left) but i don't want to mislead the user to write package imports for every mini-lib, the user should just use the big "stagexl" library.

If it is still unclear i will try to sketch it over the weekend :) Maybe the headline of this issue isn't the best, maybe it should be more like "how to structure big libraries into logical groups of classes".

Btw. clicking in the left navigation scrolls down to the one line comment on the right side. This feels somehow wrong to me. The right side with the one line comments should be an overview of available classes or methods, but when i click on a link in the left navigation i would prefer to see the detailed docs (but that's a totally different story and not related to this issue).

**Comment by sethladd** [comment]
> Btw. clicking in the left navigation scrolls down to the one line comment on the right side. 

This is now fixed. :) Clicking on the left-hand nav takes you to a dedicated page.

> The old docs show my internal libraries on the left

Ah, thanks for the clarification. So here's the tension... you want the dartdocs to show internal structure, but the user doesn't actually import that internal structure. We've built the new dartdocs to try to represent how a user would use and see a package/library. We didn't want to display internal structure, because the user doesn't actually use the internal structure.

I feel like dartdoc is doing the right thing here: documenting how a user sees and would use a library/package. We don't want to show libs from the src/ directory. Those aren't meant to be imported.

I definitely understand the desire to use the library "feature" of Dart to logically group items together. This helps with discoverability. However, if you don't export those libraries, the user can get confused if they see the library in the docs, but can't import it.

I think "Dart style" is to put your mini libs into the root of lib/ and re-export them all via your main stagexl package. We could use API docs to inform the user, if they are looking at a mini-lib, that they probably want to import all of stagexl.

**Comment by sethladd** [comment]
Here's another idea:

What if dartdoc displayed directory structure of what's inside of lib/ (and not in src/) as a way of hinting/indicating organization?

You could have:

```
lib/
  stagexl.dart
  animation/
  drawing/
  engine/
  src/
   ..implementation details..
```

Of course, this means that a user could import stagexl/animation/foo.dart because it's not in src.

**Comment by sethladd** [comment]
I took a look at how StageXL is organized: https://github.com/bp74/StageXL/tree/master/lib

I see a single library that a user imports (stagexl, inside the lib/ dir). This is what dartdoc will document and expose. We made this choice to help ensure that users are interacting with a library in the correct way.

You'll need to create user-importable libraries and put them in lib/ (or, more accurately, outside of src/) if you want to expose some sort of grouping.

Sorry about this trouble. I really appreciate that you took the time to provide feedback. I'm going to close this issue for now.
